### DISCUSSION TITLE
#ApiPlatform: how to call $iriConverter->getSubresourceIriFromResourceClass(

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#46754] [Q&A] [answered]
How do I get the route for a subresource in API Platform? I don't understand what should be in the $identifiers argument.

```php
#[ORM\Entity]
#[ApiResource]
class Question
{
    #[ORM\OneToMany(targetEntity: Answer::class, mappedBy: 'question']
    #[ApiSubresource]
    public $answers;

```

```php
$question = $questionRepo->find(42);

$iriConverter->getIriFromResourceClass($question::class); 
// returns /api/questions
$questionId = $iriConverter->getIriFromItem($question); 
// returns /api/questions/42

# desired return /api/questions/42/answers
# unsuccessful attempts:
$iriConverter->getSubresourceIriFromResourceClass($question, ['answers'] ); 
$iriConverter->getSubresourceIriFromResourceClass($question::class, ['question' => $questionId, 'property' => 'answers'] ); 
```

What are the arguments to getSubresourceIriFromResourceClass()?

**Comment by stof** [comment]
I suggest you to ask in the ApiPlatform support channels for that. See https://api-platform.com/community

**Comment by tacman** [comment]
Thanks, I have posted it to SO: https://stackoverflow.com/questions/72686881/how-do-i-generate-a-subresources-route-in-apiplatform

and also on the Symfony Slack Channel: https://symfony-devs.slack.com/archives/C3EQ7S3MJ/p1655995106283179

I've spent an embarrassing amount of time trying to track this done, walking through the source code even, but am stuck.  @dunglas , can you please point me in the right direction?

**Comment by alanpoulain** [comment, marked as answer]
If you check the test https://github.com/api-platform/core/blob/ffda414da9fd025b360fd7d5a5864fef6b5b9729/tests/Bridge/Symfony/Routing/IriConverterTest.php#L198, you can see that you need to do something like this:
```php
$iriConverter->getSubresourceIriFromResourceClass(
    $question::class,
    [
        'subresource_identifiers' => ['id' => $questionId],
        'subresource_resources' => [Answer::class => null],
    ]
); 
```
Note that in API Platform 2.7/3, the concept of subresources does not exist anymore, and you need to follow this documentation in order to do the same as before: https://api-platform.com/docs/main/core/subresources.

**Reply by tacman** [reply]
Thanks, and thanks for the heads-up about the 2.7 changes.

**Comment by tacman** [comment]
```php
$iriConverter->getSubresourceIriFromResourceClass(
    $question::class,
    [
        'subresource_identifiers' => ['id' => $questionId],
        'subresource_resources' => [Answer::class => null],
    ]
); 
```

**Comment by tacman** [comment]
Since subresources are going away, I upgraded to 2.7 beta 4.  

```php
#[ApiResource(
    uriTemplate: '/question/{questionId}/answers',
    uriVariables: [
        'questionId' => new Link(
            fromClass: Question::class,
            fromProperty: 'answers'
        )
    ],
    operations: [new GetCollection()]
)]
```

But I'm still not sure how to get the IRI:
```php
        $operation = (new GetCollection())->withClass(Answer::class);
        dd($iriConverter->getIriFromResource($question, operation:$operation));
```

When I look at the tests, the route seems to be generated with 
```php
$operation = (new GetCollection())->withClass(Dummy::class);
        $routerProphecy = $this->prophesize(RouterInterface::class);
        $routerProphecy->generate($operationName, ['id' => 1], UrlGeneratorInterface::ABS_URL)->shouldBeCalled()->willReturn('/dummies/1/foo');
```

Thanks.
